posts/2024-06-28-gno-debugger/README.md

```diff
@@ -0,0 +1,218 @@
+---
+title: "Debugging Gno Programs"
+publication_date: 2024-06-28T13:37:00Z
+slug: gno-debugger
+tags: [blog, post, tutorial, gno, debugger]
+authors: [mvertes]
+---
+# Debugging Gno Programs
+ 
+In this article, we introduce the new Gno debugger feature and show how it can be used to better understand Gno programs and help fixing bugs.
+
+## Motivation for a Gno debugger
+
+> Debugging is twice as hard as writing code.
+> Brian Kerninghan, "The Elements of Programming Style"
+
+> On average, you spend about eight to ten times debugging as you do writing code.
+> Anonymous
+
+Having a good debugger is important. But the Gno language is almost Go, and gno.land itself is entirely written in Go. Could I just use the existing Go tools, i.e.  the [delve] debugger, to take control and debug my Gno programs?
+
+You cannot debug your *Gno* program this way because doing so would entail debugging the Gno virtual machine rather than your own program. The relevant state information would be opaque and would need to be reversed and reconstructed from internal Gno virtual machine data structures.
+
+The Gno debugger addresses this issue by displaying the state of the Gno program memory symbolically. It allows for control of program execution at the source code level, regardless of the virtual machine implementation.
+
+## Setting up
+
+The Gno debugger is fully integrated in the [gno](https://docs.gno.land/gno-tooling/cli/gno-tooling-gno) single binary, which is the tool required to build, test and run gno programs locally.
+
+There is no need to install a specific tool. You just have to install the `gno` tool itself with:
+
+```shell=
```

Comment:
I'm not sure if this works on when rendering on the platform we publish on, but it doesn't work on my local preview. It does work if the equals signs are removed. Maybe remove any special symbols outside of the syntax name in all of the code blocks? I see `=` and `!` throughout the document.

Reply:
The trailing `=` and `!` at end of markdown code delimiters were inserted by the `hackmd.io` editor, which I used to edit this post. I removed them.